### PR TITLE
agent: Reduce logging

### PR DESCRIPTION
**Pull request by petuhovskiy**
Try to reduce some very common logs:
- Change interval for recurring scheduler requests (5s -> 15s)
- Log 1-2 log entries instead of 4 for each scheduler request
- Log successful healthchecks only once per 10 requests
- Do not log "Making metrics request to VM", same info as in "Updated metrics"

ref https://github.com/neondatabase/cloud/issues/15591
ref https://github.com/neondatabase/cloud/issues/15605

### CODE REVIEW

File: pkg/agent/core/state.go [hunk 1]
```diff
@@ -199,9 +206,10 @@ func (ns *neonvmState) ongoingRequest() bool {
 func NewState(vm api.VmInfo, config Config) *State {
 	return &State{
 		internal: state{
-			Config: config,
-			Debug:  false,
-			VM:     vm,
+			Config:          config,
+			Debug:           false,
+			LoggedResources: api.Resources{VCPU: 0, Mem: 0},
```
Comment (sharnoff) [inline pkg/agent/core/state.go:211]: This is a somewhat tricky change to do well, and will intersect with #983. Let's split into a separate PR?

(broader context is that `core.State` is complicated and we do want to periodically log the desired resources even if they haven't changed (either time-based or "metrics changed"), and sometimes we'll want to immediately log them even if they're equal to what the VM is using.)

File: pkg/agent/dispatcher.go [hunk 2]
```diff
@@ -165,16 +166,29 @@ func NewDispatcher(
 		var firstSequentialFailure *time.Time
 		continuedFailureAbortTimeout := time.Second * time.Duration(runner.global.config.Monitor.MaxHealthCheckSequentialFailuresSeconds)
 
+		// if we don't have any errors, we will log only every 10th successful health check
+		const maxSuccessiveSkips = 10
+		var successesSkipped int
```
Comment (sharnoff) [inline pkg/agent/dispatcher.go:171]: WDYT about not resetting this when we log, and instead just counting the number of sequential successes/failures and including the opposite in the logs?

(i.e., on first success, log the number of failures; and vice versa)

File: pkg/agent/dispatcher.go [hunk 3]
```diff
@@ -302,7 +323,12 @@ func (disp *Dispatcher) send(ctx context.Context, logger *zap.Logger, id uint64,
 	// by base64 encoding it, so use RawMessage to avoid serializing to []byte
 	// (done by SerializeMonitorMessage), and then base64 encoding again
 	raw := json.RawMessage(data)
-	logger.Info("sending message to monitor", zap.ByteString("message", raw))
+	if bytes.HasPrefix(data, []byte("{\"content\":{},\"type\":\"HealthCheck\",\"id\":")) {
+		// don't log health checks, they're too frequent
+		logger.Debug("sending message to monitor", zap.ByteString("message", raw))
+	} else {
+		logger.Info("sending message to monitor", zap.ByteString("message", raw))
+	}
```
Comment (sharnoff) [inline pkg/agent/dispatcher.go:331]: IMO it's ok to downgrade this to debug in all cases. Any messages we send that aren't health checks will already have log lines from `exec_monitor.go`

File: pkg/agent/dispatcher.go [hunk 4]
```diff
@@ -413,7 +439,13 @@ func (disp *Dispatcher) HandleMessage(
...
 		return fmt.Errorf("Error receiving message: %w", err)
 	}
-	logger.Info("(pre-decoding): received a message", zap.ByteString("message", message))
+
+	if bytes.HasPrefix(message, []byte("{\"type\":\"HealthCheck\",\"id\":")) {
+		// don't log health checks, they're too frequent
+		logger.Debug("(pre-decoding): received a message", zap.ByteString("message", message))
+	} else {
+		logger.Info("(pre-decoding): received a message", zap.ByteString("message", message))
+	}
```
Comment (sharnoff) [inline pkg/agent/dispatcher.go:448]: Similar thing here w.r.t. downgrading to "always debug"

File: pkg/agent/runner.go [hunk 5]
```diff
@@ -778,7 +779,12 @@ func (r *Runner) DoSchedulerRequest(
 	}
 	request.Header.Set("content-type", "application/json")
 
-	logger.Info("Sending request to scheduler", zap.Any("request", reqData))
+	if reqData.LastPermit != nil && *reqData.LastPermit == reqData.Resources {
+		// If the last permit is the same as the current request, we can skip request logging.
+		logger.Debug("Sending request to scheduler", zap.Any("request", reqData))
+	} else {
+		logger.Info("Sending request to scheduler", zap.Any("request", reqData))
+	}
```
Comment (sharnoff) [inline pkg/agent/runner.go:787]: I'd rather keep this as debug-only and put the conditional in `exec_plugin.go`, so that it takes place while we hold the lock (just nicer to prioritize log lines that are guaranteed to be in the same order as the operations that took place :sweat_smile:)

File: pkg/agent/runner.go [hunk 6]
```diff
@@ -807,7 +813,8 @@ func (r *Runner) DoSchedulerRequest(
 		return nil, fmt.Errorf("Bad JSON response: %w", err)
 	}
 
-	logger.Info("Received response from scheduler", zap.Any("response", respData))
+	// there will be "Plugin request successful" INFO log right after
```
Comment (sharnoff) [inline pkg/agent/runner.go:816]: I found this comment a little confusing ("where?" ; "in what cases?"). IMO it's ok to say nothing here, but adding more context about where the log line comes from would also be ok

**Review by sharnoff — COMMENTED** [review]
nice! A few comments, but broadly pretty good

File: pkg/agent/core/state.go (hunk 1 above)
Comment (petuhovskiy) [reply pkg/agent/core/state.go]: Splitted into https://github.com/neondatabase/autoscaling/pull/1014

File: pkg/agent/runner.go (hunk 5 above)
Comment (petuhovskiy) [reply pkg/agent/runner.go:786]: My idea was that `exec_plugin.go` doesn't know exact `reqBody` sent to the scheduler, so request log in `runner.go` has more context.

Also I'm not sure about the conditional, what is the best balance between logs cost / usefulness here in your opinion?

WDYT about such log levels and no conditions?
1. [info] Starting plugin request (exec_plugin.go)
2. [debug] Sending request to scheduler (runner.go)
3. [debug] Received response from scheduler (runner.go)
4. [info] Plugin request successful (exec_plugin.go)

File: pkg/agent/dispatcher.go (hunk 2 above)
Comment (petuhovskiy) [reply pkg/agent/dispatcher.go:170]: I didn't get the idea, do think just counting totals would work?

I.e. to maintain `totalSuccessCnt` and `totalFailCnt` and always add them to `logFields`?

Then print successful healthchecks only if `totalSuccessCnt % 10 == 0`?

Comment (sharnoff) [reply pkg/agent/dispatcher.go:170]: Roughly, yeah- with the additions that

1. `totalSuccessCnt` is reset to 0 on failure
2. `totalFailCnt` is reset to 0 on success
3. Failure-then-success logs `totalFailCnt`
4. Success-then-failure logs `totalSuccessCnt`

Comment (petuhovskiy) [reply pkg/agent/dispatcher.go]: Ok, made the change.

> Failure-then-success logs totalFailCnt
> Success-then-failure logs totalSuccessCnt

IMO it's not necessary to do this. I pushed a commit now which just counts oks/fails in a row.

File: pkg/agent/runner.go (hunk 5 above)
Comment (sharnoff) [reply pkg/agent/runner.go]: > WDYT about such log levels and no conditions?

That makes sense, yeah - I think we _could_ also change (1) to debug in the cases where it doesn't change, but it's easy enough to change that in a follow-up

> My idea was that exec_plugin.go doesn't know exact `reqBody` sent to the scheduler, so request log in `runner.go` has more context.

Yeah, that's what I was thinking first as well, but eventually thought it'd be better to keep the guarantees around log lines being in the right order -- especially because IIRC `reqBody` can be exactly determined from the action given to `exec_plugin.go`

**Review by sharnoff — APPROVED** [review]
lgtm, nice!
